lib/puppet/parser/functions/sprintf.rb

```diff
@@ -29,8 +29,22 @@
   :arity => -2,
   :doc => "Perform printf-style formatting of text.
 
-      The first parameter is format string describing how the rest of the parameters should be formatted.  See the documentation for the `Kernel::sprintf` function in Ruby for all the details."
+  The first parameter is format string describing how the rest of the parameters should be formatted.
+  See the documentation for the `Kernel::sprintf` function in Ruby for all the details."
 ) do |args|
-  fmt = args.shift
-  return sprintf(fmt, *args)
+  fmt = args[0]
+  args = args[1..-1]
+  begin
+    return sprintf(fmt, *args)
+  rescue KeyError => e
+    if args.size == 1 && args[0].is_a?(Hash)
+      # map the single hash argument such that all top level string keys are symbols
+      # as that allows named arguments to be used in the format string.
+      #
+      result = {}
+      args[0].each_pair { |k,v| result[k.is_a?(String) ? k.to_sym : k] = v }
```

Comment:
Why not just `result[k.to_sym] = v` here? `Symbol#to_sym` will return itself.

Reply:
Is there a point in including entries that are not converted to symbols in the `result` hash?